packages/@aws-cdk/aws-ec2/lib/cfn-init.ts

```diff
@@ -0,0 +1,362 @@
+
+    attachOptions.instanceRole.addToPolicy(new iam.PolicyStatement({
+      actions: ['cloudformation:DescribeStackResource', 'cloudformation:SignalResource'],
+      resources: [Aws.STACK_ID],
+    }));
+
+    if (bindResult.authData) {
+      attachedResource.addMetadata('AWS::CloudFormation::Authentication', bindResult.authData);
+    }
+
+    // To identify the resources that have the metadata and where the signal
+    // needs to be sent, we need { region, stackName, logicalId }
+    const resourceLocator = `--region ${Aws.REGION} --stack ${Aws.STACK_NAME} --resource ${attachedResource.logicalId}`;
+    const configSets = (attachOptions.configSets ?? ['default']).join(',');
+    const printLog = attachOptions.printLog ?? true;
+
+    if (attachOptions.embedFingerprint ?? true) {
+      // It just so happens that the comment char is '#' for both bash and PowerShell
+      attachOptions.userData.addCommands(`# fingerprint: ${fingerprint}`);
+    }
+
+    if (attachOptions.platform === InitPlatform.WINDOWS) {
+      const errCode = attachOptions.ignoreFailures ? '0' : '$LASTEXITCODE';
+      attachOptions.userData.addCommands(...[
+        `cfn-init.exe -v ${resourceLocator} -c ${configSets}`,
+        `cfn-signal.exe -e ${errCode} ${resourceLocator}`,
+        ...printLog ? ['type C:\\cfn\\log\\cfn-init.log'] : [],
+      ]);
+    } else {
+      const errCode = attachOptions.ignoreFailures ? '0' : '$?';
+      attachOptions.userData.addCommands(...[
+        // Run a subshell without 'errexit', so we can signal using the exit code of cfn-init
+        '(',
+        '  set +e',
+        `  /opt/aws/bin/cfn-init -v ${resourceLocator} -c ${configSets}`,
+        `  /opt/aws/bin/cfn-signal -e ${errCode} ${resourceLocator}`,
+        ...printLog ? ['  cat /var/log/cfn-init.log >&2'] : [],
+        ')',
+      ]);
+    }
+  }
+
+  private bind(scope: Construct, options: AttachInitOptions): { configData: any, authData: any } {
+    const nonEmptyConfigs = mapValues(this._configs, c => c.isEmpty ? undefined : c);
+
+    const configNameToBindResult = mapValues(nonEmptyConfigs, c => c._bind(scope, options));
+
+    return {
+      configData: {
+        configSets: mapValues(this._configSets, configNames => configNames.filter(name => nonEmptyConfigs[name] !== undefined)),
+        ...mapValues(configNameToBindResult, c => c.config),
+      },
+      authData: Object.values(configNameToBindResult).map(c => c.authentication).reduce(deepMerge, undefined),
+    };
+  }
+
+}
+
+/**
+ * Options for attach a CloudFormationInit to a resource
+ */
+export interface AttachInitOptions {
+  /**
+   * Instance role of the consuming instance or fleet
+   */
+  readonly instanceRole: iam.IRole;
+
+  /**
+   * OS Platfrom the init config will be used for
+   */
+  readonly platform: InitPlatform;
+
+  /**
+   * UserData to add commands to
+   */
+  readonly userData: UserData;
+
+  /**
+   * ConfigSet to activate
```

Comment:
(If this is going to be a public API) nice to either briefly describe or link to what config sets are.